src/HealthGPS/energy_balance_model.cpp

```diff
@@ -35,66 +46,163 @@ void EnergyBalanceModel::generate_risk_factors([[maybe_unused]] RuntimeContext &
             continue;
         }
 
-        auto current_risk_factors =
-            get_current_risk_factors(context.mapping(), entity, context.time_now());
-
-        // Model calibrated on previous year's age
-        auto model_age = static_cast<int>(entity.age - 1);
-        if (current_risk_factors.at(age_key) > model_age) {
-            current_risk_factors.at(age_key) = model_age;
-        }
+        auto trial = simulate_person(person, false, 0.0);
+        mean_sim_body_weight += trial[0];
+        mean_adjustment_coefficient += trial[1];
+    }
 
-        [[maybe_unused]] double energy_intake = 0.0;
-        std::unordered_map<core::Identifier, double> nutrient_intakes;
+    // Compute baseline adjustment.
+    mean_sim_body_weight /= population_size;
+    mean_adjustment_coefficient /= population_size;
+    double final_shift = (target_BW - mean_sim_body_weight) / mean_adjustment_coefficient;
 
-        // Initialise nutrients to zero.
-        for (const auto &coefficient : energy_equation_) {
-            nutrient_intakes[coefficient.first] = 0.0;
+    // Final run.
+    for (auto &person : population) {
+        // Ignore if inactive.
+        if (!person.is_active()) {
+            continue;
         }
 
-        // Compute nutrient intakes from food intakes.
-        for (const auto &equation : nutrient_equations_) {
-            // TODO: until carbs is added to risk factors (should use at, not [])
-            // double food_intake = current_risk_factors.at(equation.first);
-            double food_intake = current_risk_factors[equation.first];
+        simulate_person(person, true, final_shift);
+    }
+}
 
-            for (const auto &coefficient : equation.second) {
-                double delta_nutrient = food_intake * coefficient.second;
-                nutrient_intakes[coefficient.first] += delta_nutrient;
-            }
-        }
+std::array<double, 2> EnergyBalanceModel::simulate_person(Person &person, bool final_run,
+                                                          double final_shift) const {
+    // Model initial state.
+    const double H_0 = person.get_risk_factor_value(H_key);
+    const double BW_0 = person.get_risk_factor_value(BW_key);
+    const double PAL_0 = person.get_risk_factor_value(PAL_key);
+    const double F_0 = person.get_risk_factor_value(F_key);
+    const double L_0 = person.get_risk_factor_value(L_key);
+    const double ECF_0 = person.get_risk_factor_value(ECF_key);
+    const double EI_0 = person.get_risk_factor_value(EI_key);
+    const double CI_0 = person.get_risk_factor_value(CI_key);
+    const double G_0 = 0.5;
+
+    // TODO: Update height.
+    double H = H_0;
+
+    // TODO: Update physical activity level.
+    double PAL = PAL_0;
+
+    // Compute nutrient intakes from food intakes.
+    std::unordered_map<core::Identifier, double> nutrient_intakes;
+    for (const auto &coefficient : energy_equation_) {
+        nutrient_intakes[coefficient.first] = 0.0;
+    }
```

Comment:
STL containers default-initialise their contents anyway, so you don't need this:

```suggestion
```